### ISSUE TITLE
Semicolon woes

### ISSUE THREAD

**Issue by bitcrazed** [rust-lang/rust#79741]
## I tried this code:

```rust
fn write_db(key: String, val: String) -> Result<(), Error> {
    std::fs::write(
        "hello.txt", 
        format!("{} {}", key, val));
}
```

## I expected to see this happen: 

Compiled successfully

## Instead, this happened: 

The following error message was reported by the compiler which was far from helpful:

```
error[E0308]: mismatched types
  --> src\main.rs:15:42
   |
15 | fn write_db(key: String, val: String) -> Result<(), Error> {
   |    --------                              ^^^^^^^^^^^^^^^^^ expected enum `std::result::Result`, found `()`
   |    |
   |    implicitly returns `()` as its body has no tail or `return` expression
...
18 |         format!("{} {}", key, val));
   |                                    - help: consider removing this semicolon
   |
   = note:   expected enum `std::result::Result<(), std::io::Error>`
           found unit type `()`
```
To be fair, the compiler does hint that I should've removed the trailing `;` (semicolon), but one only sees this if one reads all the way through the entire error ... which isn't reported in full by tooling:

![image](https://user-images.githubusercontent.com/961950/101247869-b2745d00-36d0-11eb-8aa4-a196dedbbfdc.png)

So the error here isn't that there's a type mismatch; the error is that there's a trailing ; which for "reasons" returns a nil `()` expression

## A plea to the Rust team
Rust is often proposed as a "better" __replacement for__ C/C++. As such, please consider the plight of C/C++ developers learning and porting large bodies of C/C++ code to Rust.

While Rust shouldn't aim to be a better C/C++, please try to inform C/C++ developers where their C/C++-isms won't work, wherever possible. In fact, I'd argue that suggested fixes should come before explanations in error messages.

I spent 4 hours on the issue above, reading docs, SO issues, rewrote the code 4-5 times. Checked to see if I was running the latest compiler version, installed half a dozen Rust addins for VSCode to hope that one of them could tell me what I was doing wrong, etc. 

At no point did I ever consider that the error was caused by a terminating semicolon! 😠 

And it was all because Rust has a rule that a semicolon should suppress the result of an expression.

In which case ...

```rust
let foo = 2;
```
... shouldn't set `foo` to `2` because the `;` should return the nil expression, no? Isn't this an inconsistent application of the semicolon rule?

### Meta

`rustc --version --verbose`:
```
rustc 1.48.0 (7eac88abb 2020-11-16)
binary: rustc
commit-hash: 7eac88abb2e57e752f3302f02be5f3ce3d7adfb4
commit-date: 2020-11-16
host: x86_64-pc-windows-msvc
release: 1.48.0
LLVM version: 11.0
```

<details><summary>Backtrace</summary>
<p>

```
None
```

</p>
</details>

**Comment by jyn514** [comment]
@bitcrazed I'd say the bug here is that the IDE isn't reporting the full error message; if it had just shown that last line you would have seen the problem almost immediately.

**Comment by jyn514** [comment]
> let foo = 2; shouldn't set foo to 2 because the ; should return the nil expression, no? Isn't this an inconsistent application of the semicolon rule?

`let foo = 2;` is one statement, not an expression; `let foo = 2` is invalid syntax.

**Comment by SNCPlay42** [comment]
In VSCode with the rust-analyzer extension (I assume you're using the "Rust" extension?), the suggestion is also omitted from the error hover text, but *is* offered as a quick-fix. Which perhaps isn't very discoverable if you're not expecting it, but is better.

I'd consider it a bug in whichever extension you're using if it's not displaying the suggestion and just ignoring it entirely.

**Comment by SNCPlay42** [comment]
It's also unhelpful that annotations are just displayed as `file.rs (LL, CC): message` in the diagnostics's main span's hover text, requiring you to click to see the location it's referring to (and we've lost information about where these spans *end*), rather than all the annotations appearing simultaneously in context like they do in the compiler's text output. And because suggestions are not displayed (to be provided as a quick-fix or otherwise), we don't even get that for suggestions. The compiler's diagnostics tend to assume that it can annotate lots of spans of code in a convenient way and it's a real shame that the experience is worsened somewhat in IDEs.

I'm not really sure what could be done to improve that. Have each individual annotation appear as its own highlight on the actual annotated code? But that makes it difficult to determine their context. Or perhaps multiple tooltips that appear when you hover over the primary location? But that would need IDE features.

**Comment by jyn514** [comment]
cc @matklad

**Comment by SNCPlay42** [comment]
IDE issues aside, there could be value in a note like

>`this is a statement that discards its result, rather than a tail expression, because it is followed by a semicolon`

to explain *why* removing a semicolon would have an effect. This *is* mentioned (`as its body has no tail`) but it's terse and requires you to know what it means by that.

**Comment by ehuss** [comment]
FWIW, this is what it looks like in Sublime:
<img width="663" alt="image" src="https://user-images.githubusercontent.com/43198/101260669-210fe580-36e6-11eb-8585-d79946cde6a1.png">
Which I think makes the suggestion very clear.  Many people hate inline diagnostics like this, though.  VSCode (and also IntelliJ) chose a less invasive approach to showing diagnostics, which unfortunately means the entire message may not be very evident if you're not familiar with how they are displayed. I don't use VSCode, but I imagine over time people become more attuned to noticing the light bulb quick-fix.

Aside from IDE issues, I'm not sure how the diagnostic itself could be improved. Showing the lines out-of-order I think could add to confusion.  Also, the compiler doesn't know whether the function is just unfinished.  Maybe you intended to have a line afterwards like `Ok(())`, or otherwise the function is just unfinished, in which case the "no tail or `return` expression" is correct in that the return value was forgotten.

As for helping new users, I think something like `-Zteach` could help (which causes rustc to show a long-form explanation), and maybe the VSCode could have an icon or something to show that within the popup.  I'm not sure what is possible with VSCode, though, or if that would be easily discoverable.  (It's also uncertain if `-Zteach` will go anywhere.)

**Comment by jonas-schievink** [comment]
Yeah, the IDE part is a UX issue in VS Code – rust-analyzer correctly emits the suggested fix and even marks it as "preferred" if rustc says that it's machine-applicable, VS Code just does not make that prominent enough it seems.

**Comment by jonas-schievink** [comment]
rust-analyzer/rust-analyzer#6734 now works around the UX issue and emits diagnostics for *every* highlighted span in the rustc diagnostic

**Comment by camelid** [comment]
> In which case ...
> 
> ```rust
> let foo = 2;
> ```
> 
> ... shouldn't set `foo` to `2` because the `;` should return the nil expression, no? Isn't this an inconsistent application of the semicolon rule?

Note that this will set `foo` to `()`:

```rust
let foo = {2;};
```

When you write a function like:

```rust
fn write_db(key: String, val: String) -> Result<(), Error> {
    std::fs::write(
        "hello.txt", 
        format!("{} {}", key, val));
}
```

That's theoretically equivalent to:

```rust
let write_db = |key: String, val: String| -> Result<(), Error> {
    std::fs::write(
        "hello.txt", 
        format!("{} {}", key, val));
};
```

So it's not inconsistent.

**Comment by jonas-schievink** [comment]
This is now fixed, so closing
